File: frontend/donor-pwa/src/lib/api/ticket-invitations.ts
/**
 * Ticket Invitations API Client
 *
 * Provides functions for sending, resending, and validating
 * ticket invitation links, as well as registering via invitation.
 */
import apiClient from '@/lib/axios'

// ================================
// Types
// ================================

export interface InvitationSendResponse {
  invitation_id: string
  assignment_id: string
  email_address: string
  sent_at: string
  invitation_count: number
}

export interface InvitationValidateResponse {
  valid: boolean
  expired: boolean
  already_registered: boolean
  event_name: string | null
  event_date: string | null
  event_slug: string | null
  guest_name: string | null
  guest_email: string | null
  assignment_id: string | null
  /** True when the guest has a pre-created account that needs password setup. */
  needs_account_setup: boolean
  /** One-time setup token used with the setup-and-register endpoint. */
  setup_token: string | null
}

export interface InvitationSetupAndRegisterResponse {
  registration_id: string
  event_id: string
  event_slug: string
  status: string
  access_token: string
  refresh_token: string
}

export interface InvitationRegisterResponse {
  registration_id: string
  event_id: string
  event_slug: string
  status: string
}

// ================================
// API Functions
// ================================

/**
 * Send a ticket invitation email to an assigned guest.
 */
export async function sendInvitation(
  assignmentId: string,
  personalMessage?: string
) {
  const response = await apiClient.post<InvitationSendResponse>(
    `/tickets/assignments/${assignmentId}/invite`,
    { personal_message: personalMessage }
  )
  return response.data
}

/**
 * Resend an invitation email for an assignment.
 */
export async function resendInvitation(assignmentId: string) {
  const response = await apiClient.post<InvitationSendResponse>(
    `/tickets/assignments/${assignmentId}/resend-invite`
  )
  return response.data
}

/**
 * Validate an invitation token to check if it is still valid.
 */
export async function validateInvitationToken(token: string) {
  const response = await apiClient.get<InvitationValidateResponse>(
    `/invitations/${token}/validate`
  )
  return response.data
}

/**
 * Register for an event using an invitation token.
 */
export async function registerViaInvitation(
  token: string,
  data: {
    phone?: string
    meal_selection_id?: string
    custom_responses?: Record<string, string>
  }
) {
  const response = await apiClient.post<InvitationRegisterResponse>(
    `/invitations/${token}/register`,
    data
  )
  return response.data
}

/**
 * Activate a pre-created donor account and complete registration in one step.
 * Used when validate returns needs_account_setup=true.
 */
export async function setupAndRegister(
  token: string,
  data: {
    setup_token: string
    password: string
    phone?: string
    meal_selection_id?: string
    custom_responses?: Record<string, string>
  }
) {
  const response = await apiClient.post<InvitationSetupAndRegisterResponse>(
    `/invitations/${token}/setup-and-register`,
    data
  )
  return response.data
}

File: backend/app/api/v1/ticket_invitations.py
"""Invitation token validation and guest registration endpoints.

T007 — Public token validation (no auth) and authenticated registration.
"""

from __future__ import annotations

import logging
from typing import Annotated

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.middleware.auth import get_current_active_user
from app.models.user import User
from app.schemas.ticket_purchasing import (
    InvitationRegisterRequest,
    InvitationRegisterResponse,
    InvitationSetupAndRegisterRequest,
    InvitationSetupAndRegisterResponse,
    InvitationValidateResponse,
)
from app.services.ticket_invitation_service import TicketInvitationService

logger = logging.getLogger(__name__)

router = APIRouter(tags=["ticket-invitations"])


@router.get(
    "/invitations/{token}/validate",
    response_model=InvitationValidateResponse,
)
async def validate_invitation_token(
    token: str,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> InvitationValidateResponse:
    """Validate an invitation token (public, no auth required).

    Returns event details and guest information if the token is valid,
    or indicates that it is expired / already used.
    """
    return await TicketInvitationService.validate_token(db=db, token=token)


@router.post(
    "/invitations/{token}/register",
    response_model=InvitationRegisterResponse,
)
async def register_via_invitation(
    token: str,
    body: InvitationRegisterRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_active_user)],
) -> InvitationRegisterResponse:
    """Complete event registration using an invitation token.

    The authenticated user's email must match the invitation's guest email.
    """
    response = await TicketInvitationService.register_via_invitation(
        db=db,
        token=token,
        user_id=current_user.id,
        request=body,
    )
    await db.commit()
    return response


@router.post(
    "/invitations/{token}/setup-and-register",
    response_model=InvitationSetupAndRegisterResponse,
)
async def setup_and_register_via_invitation(
    token: str,
    body: InvitationSetupAndRegisterRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> InvitationSetupAndRegisterResponse:
    """Activate account and register in one step for first-time guests (public).

    Used when the invitation response indicated ``needs_account_setup=True``.
    The guest submits their chosen password alongside the setup token; this
    endpoint activates their pre-created account, completes registration, and
    returns JWT tokens so the frontend can log them in immediately.
    """
    return await TicketInvitationService.setup_and_register(
        db=db,
        token=token,
        request=body,
    )
